RecoTracker/TkSeedGenerator/plugins/SeedingOTEDProducer.cc

```diff
@@ -0,0 +1,374 @@
+  edm::ESGetToken<MagneticField, IdealMagneticFieldRecord> magFieldToken_;
+  edm::ESGetToken<TrajectoryStateUpdator, TrackingComponentsRecord> updatorToken_;
+  edm::ESGetToken<MeasurementTracker, CkfComponentsRecord> measurementTrackerToken_;
+  edm::ESGetToken<Chi2MeasurementEstimatorBase, TrackingComponentsRecord> estToken_;
+  edm::EDPutTokenT<TrajectorySeedCollection> putToken_;
+};
+
+SeedingOTEDProducer::SeedingOTEDProducer(edm::ParameterSet const& conf)
+    : tkMeasEventToken_(consumes<MeasurementTrackerEvent>(conf.getParameter<edm::InputTag>("trackerEvent"))),
+      topoToken_(esConsumes()),
+      propagatorToken_(esConsumes(edm::ESInputTag("", "PropagatorWithMaterial"))),
+      magFieldToken_(esConsumes()),
+      updatorToken_(esConsumes(edm::ESInputTag("", "KFUpdator"))),
+      measurementTrackerToken_(esConsumes()),
+      estToken_(esConsumes(edm::ESInputTag("", "Chi2"))) {
+  vhProducerToken_ = consumes<VectorHitCollection>(edm::InputTag(conf.getParameter<edm::InputTag>("src")));
+  beamSpotToken_ = consumes<reco::BeamSpot>(conf.getParameter<edm::InputTag>("beamSpotLabel"));
+  updatorName_ = conf.getParameter<std::string>("updator");
+  putToken_ = produces<TrajectorySeedCollection>();
+}
+
+SeedingOTEDProducer::~SeedingOTEDProducer() {}
+
+void SeedingOTEDProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("src", edm::InputTag("siPhase2VectorHits", "accepted"));
+  desc.add<edm::InputTag>("trackerEvent", edm::InputTag("MeasurementTrackerEvent"));
+  desc.add<edm::InputTag>("beamSpotLabel", edm::InputTag("offlineBeamSpot"));
+  desc.add<std::string>("updator", std::string("KFUpdator"));
+  descriptions.add("SeedingOTEDProducer", desc);
+}
+
+void SeedingOTEDProducer::produce(edm::Event& event, const edm::EventSetup& es) {
+  tkTopo_ = &es.getData(topoToken_);
+
+  edm::ESHandle<MeasurementTracker> measurementTrackerHandle = es.getHandle(measurementTrackerToken_);
+  measurementTracker_ = measurementTrackerHandle.product();
+
+  auto const& measurementTrackerEvent = event.get(tkMeasEventToken_);
+
+  layerMeasurements_ = std::make_unique<LayerMeasurements>(*measurementTrackerHandle, measurementTrackerEvent);
+
+  estimator_ = &es.getData(estToken_);
+
+  propagator_ = &es.getData(propagatorToken_);
+
+  magField_ = &es.getData(magFieldToken_);
+
+  updator_ = &es.getData(updatorToken_);
+
+  beamSpot_ = &event.get(beamSpotToken_);
+
+  // Get the vector hits
+  auto vhs = event.getHandle(vhProducerToken_);
+
+  event.emplace(putToken_, run(vhs));
+}
+
+TrajectorySeedCollection SeedingOTEDProducer::run(edm::Handle<VectorHitCollection> vHs) {
+  TrajectorySeedCollection result;
+  //check if all the first three layers have VHs
+  std::vector<const VectorHit*> vhSeedsL1 = collectVHsOnLayer(*vHs.product(), 1);
+  std::vector<const VectorHit*> vhSeedsL2 = collectVHsOnLayer(*vHs.product(), 2);
```

Comment:
I realize now that both of these vectors are not used (which would be the "accepted" VH passing the cuts).
Do I understand correctly that then below we get all ("accepted"+"rejected") hits combined to build the full seed.
I think I was told that the accepted ones define the seed, I guess it was just partial.
I suppose looking at all possible combinations would further improve the efficiency, but it's likely to be relatively minor.